src/gpu/ForwardProjectorByBinNiftyPET.cxx

```diff
@@ -0,0 +1,190 @@
+//       error();
+
+    viewgrams = _projected_data_sptr->get_related_viewgrams(
+        viewgrams.get_basic_view_segment_num(), _symmetries_sptr);
+}
+
+void
+ForwardProjectorByBinNiftyPET::
+set_input(const shared_ptr<DiscretisedDensity<3,float> >& density_sptr)
+{
+    _density_sptr.reset(density_sptr->clone());
+
+    // Get dimensions of sinogram
+    int num_sinograms = _projected_data_sptr->get_num_sinograms();
+    int num_views     = _projected_data_sptr->get_num_views();
+    int num_tang_poss = _projected_data_sptr->get_num_tangential_poss();
+    int num_proj_data_elems = num_sinograms * num_views * num_tang_poss;
+    // Create array for sinogram and fill it
+    float *proj_data_ptr = new float[num_proj_data_elems];
+    // Necessary?
+    for (int i=0; i<num_proj_data_elems; ++i)
+        proj_data_ptr[i] = 0.F;
+
+    // Get dimensions of image
+    int dim[3];
+    Coordinate3D<int> min_indices;
+    Coordinate3D<int> max_indices;
+    if (!_density_sptr->get_regular_range(min_indices, max_indices))
+        throw std::runtime_error("ForwardProjectorByBinNiftyPET::set_input - "
+                                 "expected image to have regular range.");
+    for (int i=0; i<3; ++i)
+        dim[i] = max_indices[i + 1] - min_indices[i + 1] + 1;
+    // Create array for image and fill it - think carefully about index order!
+    float *im_ptr = new float[dim[0]*dim[1]*dim[2]];
+    // TODO: assert((y,x,z) = (320,320,128)) as required by NiftyPET by cropping/padding original 344x344x127
+    for (int z = min_indices[1], i = 0; z <= max_indices[1]; z++)
+		for (int y = min_indices[2]; y <= max_indices[2]; y++)
+			for (int x = min_indices[3]; x <= max_indices[3]; x++, i++)
+				im_ptr[i] = (*_density_sptr)[z][y][x];
+
+    // Probably not necessary - delete after development if not used
+    int num_segments  = this->_projected_data_sptr->get_num_segments();
+
+    Cnst Cnt;
+    Cnt.SPN = const_cast<char>(11);
+    Cnt.RNG_STRT = const_cast<char>(0);
+    Cnt.RNG_END = const_cast<char>(64);
+    Cnt.VERBOSE = false;
+    Cnt.DEVID = const_cast<char>(0);
+    Cnt.NSN11 = 837;
+    Cnt.NSEG0 = 127;
+
+    int nsinos;
+    switch(Cnt.SPN){
+      case 11:
+        nsinos = Cnt.NSN11; break;
+      case 0:
+        nsinos = Cnt.NSEG0; break;
+      default:
+        throw runtime_error("Unsupported span");
+    }
+
+    std::vector<int> isub();  // TODO: expose as argument?
+    if (isub.size() == 0) {
+      // AW #defined to be 68516, number of active bins in 2D sino
+      isub = std::vector<int>(AW);
+      for (int i = 0; i<AW; i++) isub[i] = i;
+    }
+    std::vector<int> sinog(isub.size() * nsinos, 0);
+
+    // TODO: fill in precomputed values
+    float * li2rng;    // axLUT["li2rng"]
+    short * li2sn;     // axLUT["li2sn1" if Cnt.SPN == 1 else "li2sn"]
+    char * li2nos;     // axLUT["li2nos"]
+    short *s2c;        // txLUT["s2c"]
+    int *aw2ali;       // txLUT["aw2ali"]
+    float *crss;       // txLUT["crs"]
```

Comment:
the 6 missing arguments are arrays of thousands of elements which are pre-computed by a few hundred lines of python code (vis. `niftypet.nipet.get_mmrparams()`).
As an initial hack, I'd be tempted to hardcode them in C++ syntax in a massive header.